File: src/py/models/factory.py
"""Model factory implementation."""

from typing import Any, Dict
from ..model import Model


class ModelFactory:
    """Factory class for creating model instances."""
    
    @staticmethod
    def create(config: Dict[str, Any]):
        """Create model instance based on configuration.
        
        Args:
            config: Model configuration
            
        Returns:
            Model: Model instance
        """
        model_type = config.get('type')
        
        if model_type == 'dft':
            from .dft import DFTModel
            return DFTModel(config)
        else:
            return Model(**config)
    
    @staticmethod
    def create_from_application(config: Dict[str, Any]):
        """Create model from application configuration.
        
        Args:
            config: Configuration with application info
            
        Returns:
            Model: Model instance
            
        Raises:
            ValueError: If model type cannot be determined
        """
        application = config.get('application')
        if not application:
            raise ValueError("ModelFactory.create_from_application: application is required")
        
        from ..tree import get_default_model_type_for_application
        model_type = get_default_model_type_for_application(application)
        
        if not model_type:
            raise ValueError(f"ModelFactory.create_from_application: cannot determine model type: {model_type}")
        
        return ModelFactory.create({**config, 'type': model_type})

File: src/js/models/factory.ts
import type { ApplicationSchema } from "@mat3ra/esse/dist/js/types";

import { Model } from "../model";
import {
    getDefaultModelTypeSubtypeForApplication,
    getTreeByApplicationNameAndVersion,
} from "../tree";
import type { ModelConfig } from "../types";
import { DFTModel } from "./dft";

export type ModelConfigFromApplication = Partial<ModelConfig> & { application: ApplicationSchema };

export class ModelFactory {
    static DFTModel = DFTModel;

    static Model = Model;

    static create(config: ModelConfig): Model {
        switch (config.type) {
            case "dft":
                return new this.DFTModel(config);
            default:
                return new this.Model(config);
        }
    }

    static createFromApplication(config: ModelConfigFromApplication): Model {
        const { application } = config;
        if (!application) {
            throw new Error("ModelFactory.createFromApplication: application is required");
        }

        const tree = getTreeByApplicationNameAndVersion(application);
        if (!tree || Object.keys(tree).length === 0) {
            return this.create({ ...config, type: "unknown", subtype: "unknown" });
        }

        const typeSubtype = getDefaultModelTypeSubtypeForApplication(application);

        return this.create({ ...config, ...typeSubtype });
    }
}
